crates/ironrdp-server/src/server.rs

```diff
@@ -274,26 +307,39 @@ impl RdpServer {
         }
     }
 
+    async fn handle_io_channel_data(&mut self, data: SendDataRequest<'_>) -> Result<bool> {
+        let control = rdp::headers::ShareControlHeader::from_buffer(Cursor::new(data.user_data))?;
+
+        match control.share_control_pdu {
+            rdp::headers::ShareControlPdu::Data(header) => match header.share_data_pdu {
+                rdp::headers::ShareDataPdu::Input(pdu) => {
+                    self.handle_input_event(pdu).await;
+                }
+
+                rdp::headers::ShareDataPdu::ShutdownRequest => {
+                    return Ok(true);
+                }
+
+                unexpected => {
+                    warn!(?unexpected, "Unexpected share data pdu");
+                }
+            },
+
+            unexpected => {
+                warn!(?unexpected, "Unexpected share control");
+            }
+        }
+
+        Ok(false)
+    }
+
     async fn handle_x224(&mut self, frame: &[u8]) -> Result<bool> {
         let message = ironrdp_pdu::decode::<mcs::McsMessage<'_>>(frame)?;
         match message {
             mcs::McsMessage::SendDataRequest(data) => {
-                let control = rdp::headers::ShareControlHeader::from_buffer(Cursor::new(data.user_data))?;
-
-                match control.share_control_pdu {
-                    rdp::headers::ShareControlPdu::Data(header) => match header.share_data_pdu {
-                        rdp::headers::ShareDataPdu::Input(pdu) => {
-                            self.handle_input_event(pdu).await;
-                        }
-
-                        unexpected => {
-                            warn!(?unexpected, "Unexpected share data pdu");
-                        }
-                    },
-
-                    unexpected => {
-                        warn!(?unexpected, "Unexpected share control");
-                    }
+                debug!(?data, "McsMessage::SendDataRequest");
+                if Some(data.channel_id) == self.io_channel_id && self.handle_io_channel_data(data).await? {
+                    return Ok(true);
                 }
```

Comment:
Suggestion: a more idiomatic approach would be to call `handle_io_channel_data` inside the if body and avoid nesting control flow inside the condition expression itself.
```suggestion
                if Some(data.channel_id) == self.io_channel_id {
                    return self.handle_io_channel_data(data).await;
                }
```